internal/ingest/ingest.go

```diff
@@ -609,11 +617,21 @@ func (ing *Ingester) runIngestStep(syncFinishedEvent legs.SyncFinished) {
 		if _, ok := ing.providersBeingProcessed[p]; !ok {
 			ing.providersBeingProcessed[p] = &sync.Mutex{}
 		}
+		if _, ok := ing.providerAdChainStaging[p]; !ok {
+			ing.providerAdChainStaging[p] = &atomic.Value{}
+		}
 		ing.providersBeingProcessedMu.Unlock()
-		ing.toWorkers <- toWorkerMsg{
+
+		oldAssignment := ing.providerAdChainStaging[p].Swap(workerAssignment{
```

Comment:
Does this also need to be done while the `ing.providersBeingProcessedMu` lock is held?  Otherwise, there could be a read/write race with `ing.providerAdChainStaging[p] = &atomic.Value{}` on line 621.

Reply:
621 only inserts an atomic value if there was no value set in the map. Once you have the atomic value you don't need to lock the map anymore since you are only dealing with references to the atomic value.

In other words we only set `ing.providerAdChainStaging[p] = &atomic.Value{}` once in the lock. Outside the lock we only deal with `*atomic.Value` so no need for a lock.